core/src/main/java/org/elasticsearch/common/xcontent/NamedXContentRegistry.java

```diff
@@ -0,0 +1,150 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.common.xcontent;
+
+import org.elasticsearch.ElasticsearchException;
+import org.elasticsearch.common.ParseField;
+import org.elasticsearch.common.ParseFieldMatcher;
+import org.elasticsearch.common.ParseFieldMatcherSupplier;
+import org.elasticsearch.common.ParsingException;
+import org.elasticsearch.common.xcontent.support.DelegatingXContentParser;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+import static java.util.Collections.unmodifiableMap;
+
+/**
+ * Registry of "named" XContent parsers that can read using {@link XContentParser#namedObject(Class, String, ParseFieldMatcherSupplier)}.
+ */
+public class NamedXContentRegistry {
+    /**
+     * Parses an object with the type {@code T} from parser.
+     */
+    public interface FromXContent<T, C> {
+        T fromXContent(XContentParser parser, C context) throws IOException;
+    }
+
+    /**
+     * Entry that will be added to the registry.
+     */
+    public static class Entry {
+        /** The class that this entry can read. */
+        public final Class<?> categoryClass;
+
+        /** A name for the entry which is unique within the {@link #categoryClass}. */
+        public final ParseField name;
+
+        /** A parser for which returns subclasses of {@link #categoryClass}. */
+        public final FromXContent<?, ?> fromXContent;
+
+        /** Creates a new entry which can be stored by the registry. */
+        public <T> Entry(Class<T> categoryClass, ParseField name, FromXContent<? extends T, ?> fromXContent) {
+            this.categoryClass = Objects.requireNonNull(categoryClass);
+            this.name = Objects.requireNonNull(name);
+            this.fromXContent = Objects.requireNonNull(fromXContent);
+        }
+    }
+    
+    private final Map<Class<?>, Map<String, Entry>> registry;
+
+    /**
+     * Build the registry from a list of the entries that should be in it.
+     */
+    public NamedXContentRegistry(List<Entry> entries) {
+        Map<Class<?>, Map<String, Entry>> registry = new HashMap<>();
+        for (Entry entry : entries) {
```

Comment:
Can you make this algorithm the same as that in the NamedWriteableRegistry? I think it is cleaner (does not require replacing the inner maps to make them unmodifiable.